---
title: "LDA on SIRS data"
author: "Pratheepa Jeganathan, Susan Holmes"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  BiocStyle::html_document
params:
  K: 4
  R: 500
---

# Outline
metagenomics shotgun sequencing data of ten healthy subjects before (three time-points) and after teeth flossing (six to seven time-points.) A total of 90 plasma specimens from ten subjects were extracted alongside five to six negative controls per subject and sequenced using HiSeq.Adapter-trimmed, quality-filtered, and human-subtracted reads were Kraken-classified to 1645 species. 

# Set-up

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
  message = FALSE, 
  warning = FALSE,
  fig.width = 12,
  fig.height = 8)
```

```{r read_arg}
K = params$K
K
iter = params$R
iter
```

Loading packages
```{r}
library(phyloseq)
library(dplyr)
library(magrittr)
library(rstan)
library(RColorBrewer)
library(wesanderson)
library(ggheatmap)# Github repo: #devtools::install_github("jakobbossek/ggheatmap")
library(reshape2)
library(DESeq2)
#library(Rcpp)
library(ape)
```

# Analysis

## Read phyloseq and exploratory analysis

```{r }
ps = readRDS("./Data/pshf.rds")
ps
if(dim(otu_table(ps))[1]!=ntaxa(ps)){otu_table(ps) <- t(otu_table(ps))}
```

## Adding blocks
```{r}
set1 = c("FL002","FL003","FL004","FL005")
set2 = c("FL006","FL007","FL008","FL009","FL010","FL011")

#the following negative controls are in set 2 only.
nctrl = c("Water.Keck.1", "Phix.1", "Phix.2", "Water.ctrl.1", "Water.ctrl.2")

ext.num = sample_data(ps)$SubjectName
blocks = ifelse(ext.num %in% set1, "Set1", "Set2")
sample_data(ps)$block = blocks

ps = subset_samples(ps,Sample_Type %in% c("Plasma", "Control"))
```

Identify the species that are not present in at least one `Sample_Type == Plasma` sample and removed them from the phyloseq object. Label these species as contaminants. 

```{r filter_taxa}
ps = prune_taxa(taxa_sums(ps) > 0, ps)
ps.standard = subset_samples(ps, Sample_Type %in% c("Plasma"))
prevTaxaP = apply(otu_table(ps.standard), 1, function(x){sum(x>0)})

Contaminants1 = names(prevTaxaP)[prevTaxaP == 0]
length(Contaminants1)
```

```{r}
ps = prune_taxa(prevTaxaP > 0, ps)
ps
```

We identifed 255 species not in any plasma samples and they are classified as contaminants before using BARBI LDA (VB).
  
See a summary of the `plasma` and `control` samples in each block. 

```{r summary_stat}
table(sample_data(ps)$Sample_Type, sample_data(ps)$block)
```


## Prepare the phyloseq object for the Bayesian inference

We use Bayesian inference to identify contaminants in each block separately to account for the batch-effects of contamination. 

Thus, split the phyloseq object into multiple phyloseq objects corresponding to each block, and store the phyloseq objects as a list of phyloseq objects, `psByBlock`. 

Select negative control samples from each block and store as a list of phyloseq objects, `psNCbyBlock`. 

Select all species that have a prevalence of zero (i.e., have zero reads) in all negative control samples for each block and store as a list of phyloseq objects, `psallzeroInNC`.

Select all plasma samples from each block and store as a list of phyloseq objects, `psPlByBlock`.

```{r list_of_phyloseq}
source("psBlockResults.R")
psBlockResult = psBlockResults(ps, 
                               sampleTypeVar = "Sample_Type", 
                               caselevels = c("Plasma"), 
                               controllevel="Control", 
                               sampleName = "SampleCode", 
                               blockVar = "block")

psByBlock = psBlockResult[[1]]
psNCbyBlock = psBlockResult[[2]]
psallzeroInNC = psBlockResult[[3]]
psPlByBlock = psBlockResult[[4]]
```

### LDA 

Each sample is modeled as having a mixture of topics, with each species drawn from a topic based on mixing proportions.


### LDA on plasma and negative control samples


```{r}
f = stan_model(file = "lda.stan")
```

For each block, make a stan data'

```{r}
K = as.list(c(4,3)) # number of topics
stan.data.all.blocks = lapply(as.list(1:2), function(i){
  y = psByBlock[[i]]
  x = t(get_taxa(y))
  dimnames(x) = NULL
  stan_data = list(K = K[[i]], 
    V = ncol(x), 
    D = nrow(x), 
    n = x, 
    alpha = rep(1, K[[i]]), 
    gamma = rep(0.5, ncol(x)))
  return(stan_data)
})
```

```{r}
stan.fit.all.blocks = lapply(stan.data.all.blocks, function(u){
  stan_fit = sampling(f, 
                      data = u, 
                      iter = iter, 
                      chains = 2, 
                      sample_file = "./Results/sample_file_HF",
                      diagnostic_file = "./Results/diagnostic_file_HF",
                      cores = 2)
  
  return(stan_fit)
})

saveRDS(stan.fit.all.blocks, file = "./Results/SIRS_HF.rds")
```

```{r}
stan.fit.all.blocks = readRDS("./Results/SIRS_HF.rds")
```

### Extract posterior samples
```{r}
samples.all.blocks = lapply(stan.fit.all.blocks, function(x){
  rstan::extract(x, permuted = TRUE, inc_warmup = FALSE, include = TRUE)
})
```

### Posterior distirbution of topic in each sample in each block
```{r}
theta.all.blocks = lapply(as.list(1:2), function(i){
  x = samples.all.blocks[[i]]
  theta = x$theta # chain * sample * topic
  # find theta distirbution of each topic in each sample
  names(theta) = c("theta", "Sample", "Topic")
  dimnames(theta)[[2]] = sample_names(psByBlock[[i]])
  dimnames(theta)[[3]] = c(paste0("Bacterial Community ", seq(1, K[[i]])))
  theta.all = melt(theta)
  colnames(theta.all) = c("Chain", "Sample", "Topic", "topic.dis")

  theta.all$Sample = factor(theta.all$Sample)
  theta.all$Topic = factor(theta.all$Topic)

  # add control or dilution series
  sam = sample_data(psByBlock[[i]]) %>% data.frame()
  sam$unique_names = rownames(sam)
  theta.all = left_join(theta.all, sam, by =c("Sample"= "unique_names"))
  return(theta.all)

})

```


### boxplot 
```{r}

p.all.blocks = lapply(theta.all.blocks, function(u){
  sel.samples = c(as.character(unique(u$Sample[(u$Sample_Type == "Control")])), 
                  as.character(unique(u$Sample[(u$Sample_Type == "Plasma")])[10:19]))
  u = filter(u, Sample %in% sel.samples)
  p = ggplot(data = u) + 
  geom_boxplot(aes(x = Sample, y = topic.dis, color = Topic)) + 
  facet_grid(Topic ~ Sample_Type, scales = "free_x")+
  ylab(bquote(theta[k])) + 
  ggtitle("") + 
  theme(plot.title = element_text(hjust = 0.5, size = 15), 
    strip.text.y= element_text(size = 8), 
    strip.text.x = element_text(size = 12),
    axis.title.y = element_text(size = 12),
    legend.position = "none",
    axis.text.x = element_text(angle = 90, hjust = 1))
  
  return(p)
})

```

```{r fig.width=15, fig.height=10}
p.all.blocks[[1]]
ggsave("./Figures/HF_LDA_theta_set1.eps", plot = p.all.blocks[[1]], width = 9, height = 10)
```

```{r fig.width=15, fig.height=10}
p.all.blocks[[2]]
ggsave("./Figures/HF_LDA_theta_set2.eps", plot = p.all.blocks[[2]], width = 9, height = 10)
```

### ASV cloud in each topic
```{r}
beta.all.blocks = lapply(as.list(1:2), function(i){
  x = samples.all.blocks[[i]]
  beta = x$beta
  dimnames(beta)[[2]] = c(paste0("Bacterial Community ", seq(1,K[[i]])))
  
  tax_tab = tax_table(psByBlock[[i]]) %>% data.frame()
  tax_tab = mutate(tax_tab, seq.variant = rownames(tax_tab))

  dimnames(beta)[[3]] =tax_tab[, "seq.variant"]
  beta.all = melt(beta)
  colnames(beta.all) = c("Chain", "Topic", "ASV", "ASV.distribution")
  beta.all = left_join(beta.all, tax_tab, by = c("ASV"= "seq.variant"))
  beta.all$Topic = factor(beta.all$Topic)
  beta.all$ASV = factor(beta.all$ASV)
  return(beta.all)
})

```

```{r}
library(ggwordcloud)
max.beta.in.each.asv.all.topics.all.blocks = lapply(as.list(1:2), function(i){
  max.beta.in.each.asv.all.topics = group_by(beta.all.blocks[[i]], Topic, Species) %>% summarise(max_beta = max(ASV.distribution)) %>% top_n(10, max_beta) %>% as.data.frame()
  
  return(max.beta.in.each.asv.all.topics)
})

max.beta.in.each.asv.all.topics.all.blocks = lapply(max.beta.in.each.asv.all.topics.all.blocks, function(x){
  x$Species = substring(x$Species, 4) %>% as.factor()
  return(x)
})
```


```{r}
p_cloud_all_blocks = lapply(max.beta.in.each.asv.all.topics.all.blocks, function(x){
  p_cloud = ggplot(x, aes(label = Species, size = max_beta, color = Species)) + 
  geom_text_wordcloud() +
  theme_minimal() +
  scale_size_area(max_size = 10) + 
  facet_wrap(~ Topic) + 
  theme(strip.text.x = element_text(size = 12, face = "bold"))

return(p_cloud)
})

```

```{r fig.width=15, fig.height=10}
p_cloud_all_blocks[[1]]

ggsave("./Figures/HF_Set1_ASV.png", p_cloud_all_blocks[[1]], width = 15, height = 10, units = "in", dpi = 300)
```


```{r fig.width=15, fig.height=10}
p_cloud_all_blocks[[2]]

ggsave("./Figures/HF_Set2_ASV.png", p_cloud_all_blocks[[2]], width = 15, height = 10, units = "in", dpi = 300)
```