---
title: "Zymbo Mock Community Data Analysis (16S rRNA gene amplicon sequencing) using LDA-STAN (Variational Bayes)"
author: "Pratheepa Jeganathan"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    toc: true
---

We have 10 negative control samples and serial dilution of ZymoBIOMICS® Microbial Community Standard.

Looking at the DADA2 pipeline ASVs 2, 3, 4, 5, 6, 7, 8, 11, 17, 37, 22 are in the ZymoBIOMICS® Microbial Community Standard.


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE, fig.width = 18, fig.height = 12)
```

Loading packages
```{r}
library(phyloseq)
library(ggplot2)
library(magrittr)
library(dplyr)
library(reshape2)
library(rstan)
```

```{r}
set.theme = theme_update(panel.border = element_blank(),
                    panel.grid = element_line(size = .8),
                    axis.ticks = element_blank(),
                    legend.title = element_text(size = 8),
                    legend.text = element_text(size = 8),
                    axis.text = element_text(size = 8),
                    axis.title = element_text(size = 8),
                    strip.background = element_blank(),
                    strip.text = element_text(size = 8),
                    legend.key = element_blank(),
    plot.title = element_text(hjust = 0.5))
```

```{r}
theme_set(theme_bw())
```

Set the computational resources

```{r}
ncores = as.integer(Sys.getenv("SLURM_NTASKS"))
if(is.na(ncores)) ncores = parallel::detectCores()
ncores
```

### Read phyloseq and exploratory analysis

```{r}
ps = readRDS("./Data/ps_zymo.rds")
if(dim(otu_table(ps))[1]!=ntaxa(ps)){
  otu_table(ps) = t(otu_table(ps))
  }
```

change order of sample names
```{r}
ncont = paste0("NegativeControl.", seq(1, 10))
stan = paste0("Standard.Dilution.1.",c(1, 6, 36, 216, 1296, 7776, 46656, 279936))

sample_data(ps)$Name = factor(sample_data(ps)$Name, levels = c(ncont,stan))

sample_names(ps) = as.character(sample_data(ps)$Name)
```

Store sequence variants, ASVs, ASV.genus, ASV.genus.species
```{r}
ASV = as.character(paste0("ASV_", seq(1,ntaxa(ps))))

ASV.Genus = paste0("ASV_", seq(1,ntaxa(ps)), "_", as.character(tax_table(ps)[,6]))

ASV.Genus.Species = paste0(ASV, "_", as.character(tax_table(ps)[,6]), "_", as.character(tax_table(ps)[,7]))

df.ASV = data.frame(seq.variant = taxa_names(ps), ASV = ASV, ASV.Genus = ASV.Genus, ASV.Genus.Species = ASV.Genus.Species)
```


```{r }
taxa_names(ps) = df.ASV$ASV.Genus.Species
```



### Microbial composition of the ZymoBIOMICS® Microbial Community Standard (Zymo Research, Irvine, CA) measured 

True ASVs at genus level

```{r}
# true.com.ASV = taxa_names(ps)[which(tax_table(ps)[,6] %in% c("Bacillus","Listeria","Staphylococcus", "Lactobacillus", "Escherichia/Shigella", "Enterococcus", "Salmonella", "Pseudomonas"))]
true.com.ASV = taxa_names(ps)[c(2:8,11,17,37,22)]
ps.true = transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
ps.true = prune_taxa(true.com.ASV, ps.true)
ps.true
```


### Create a list of true ASVs in each diluted sample

```{r}
true.com.ASV = taxa_names(ps)[c(2:8,11,17,22,37)]
# true.com.ASV = taxa_names(ps)[which((tax_table(ps)[,6] %in% c("Bacillus","Listeria","Staphylococcus", "Lactobacillus", "Escherichia/Shigella", "Enterococcus", "Salmonella", "Pseudomonas")))]

# true.com.ASV = taxa_names(ps)[which((tax_table(ps)[,6] %in% c("Bacillus","Listeria","Staphylococcus", "Lactobacillus", "Escherichia/Shigella", "Enterococcus", "Salmonella", "Pseudomonas"))&(!is.na(tax_table(ps)[,7])))]

diluted.sample.names = sample_names(ps)[3:10]

trueSeq = function(sample.name, true.com.ASV){
  df.sample.i = data.frame(ot = otu_table(ps)[, sample.name])
  names(df.sample.i) = "ot"
  df.sample.i = mutate(df.sample.i, ASV = rownames(df.sample.i))
  df.sample.i.true.seq = filter(df.sample.i, (ot > 0) & (as.character(ASV) %in% true.com.ASV))
  true.seq.sample.i = as.character(df.sample.i.true.seq$ASV)
  return(true.seq.sample.i)
}


true.seq.all.samples = lapply(diluted.sample.names, FUN = trueSeq, true.com.ASV = true.com.ASV)

##true.seq.all.samples
```

Number of true ASVs in each dilution series sample
```{r}
number.of.true.ASVs = lapply(true.seq.all.samples, function(x) {length(x)})
```

Number of unique true ASV in all samples
```{r}
number.of.unique.ASVs.in.all.samples = length(unique(unlist(true.seq.all.samples)))
```


### Check the percentage of eleven ASVs in each dilution series sample

```{r}
ps.prop = transform_sample_counts(ps, function(otu) otu/sum(otu)*100)
ps.true = prune_taxa(unique(unlist(true.seq.all.samples)), ps)
ps.prop.true = prune_taxa(taxa_names(ps.true), ps.prop)
dt = otu_table(subset_samples(ps.prop.true, 
  SampleType == "Standard")) %>% data.frame()
colnames(dt) = sample_data(subset_samples(ps.prop.true, 
  SampleType == "Standard"))$Name
#rownames(dt) = df.ASV$ASV.Genus[which(as.character(df.ASV$ASV.Genus.Species) %in% rownames(dt))]
rownames(dt) = df.ASV$ASV.Genus.Species[which(as.character(df.ASV$ASV.Genus.Species) %in% rownames(dt))]
# library(knitr)
# kable(dt)
```

### Adding blocks

We have only one block of samples.

```{r adding_blocks}
blocks = rep("Set1", nsamples(ps))

sample_data(ps)$block = blocks
```

Identify the species that are not present in at least one `SampleType == Standard` sample and removed them from the phyloseq object. Label these species as contaminants. 

```{r filter_taxa}
ps = prune_taxa(taxa_sums(ps) > 0, ps)
ps.standard = subset_samples(ps, SampleType %in% c("Standard"))
prevTaxaP = apply(otu_table(ps.standard), 1, function(x){sum(x>0)})

Contaminants1 = names(prevTaxaP)[prevTaxaP == 0]
length(Contaminants1)
```

```{r}
ps = prune_taxa(prevTaxaP > 0, ps)
ps
```

We identifed 142 ASVs not is any dilution series samples and they are classified as contaminants before using BARBI LDA (VB).
  
See a summary of the `Standard` and `Negative` samples in each block. 

```{r summary_stat}
table(sample_data(ps)$SampleType, sample_data(ps)$block)
```

```{r}
colSums(otu_table(ps))
```

###      Prepare the phyloseq object for the Bayesian inference

We use Bayesian inference to identify contaminants in each block separately to account for the batch-effects of contamination. 

Thus, split the phyloseq object into multiple phyloseq objects corresponding to each block, and store the phyloseq objects as a list of phyloseq objects, `psByBlock`. 

Select negative control samples from each block and store as a list of phyloseq objects, `psNCbyBlock`. 

Select all species that have a prevalence of zero (i.e., have zero reads) in all negative control samples for each block and store as a list of phyloseq objects, `psallzeroInNC`.

Select all plasma samples from each block and store as a list of phyloseq objects, `psPlByBlock`.

```{r list_of_phyloseq}
source("psBlockResults.R")
psBlockResult = psBlockResults(ps, 
  sampleTypeVar = "SampleType", caselevels = c("Standard"), controllevel="Negative", sampleName = "Name", blockVar = "block")

psByBlock = psBlockResult[[1]]
psNCbyBlock = psBlockResult[[2]]
psallzeroInNC = psBlockResult[[3]]
psPlByBlock = psBlockResult[[4]]

# st = sample_names(psByBlock[[1]])[-c(4:10)]
# psByBlock[[1]] = subset_samples(psByBlock[[1]], SampleCode %in% st)
```

## LDA using VB

Each sample is modeled as having a mixture of topics, with each ASV drawn from a topic based on mixing proportions.

### LDA on negative controls

```{r}
ps.c = psNCbyBlock[[1]] # all controls
x <- t(get_taxa(ps.c))
K = 10 # number of topics
dimnames(x) <- NULL
stan_data <- list(
  K = K,
  V = ncol(x),
  D = nrow(x),
  n = x,
  alpha = rep(1, K),
  gamma = rep(0.5, ncol(x))
)
```


```{r}
start_fit <- Sys.time()

f <- stan_model(file = "lda.stan")

stan_fit_c <- vb(
  f,
  data = stan_data,
  output_samples = 1000,
  eta = 1,
  adapt_engaged = FALSE
)

cat(sprintf(
  "Finished in %f minutes\n",
  Sys.time() - start_fit, 4)
)

saveRDS(stan_fit_c, file = "stan_vb_fit_c.rds")
```


```{r}
samples.c <- rstan::extract(stan_fit_c)
```

### Posterior distirbution of beta in controls (beta for each topic - not for each topic in each sample)

```{r}
beta.c = samples.c$beta

mean.beta.c = apply(beta.c, c(2,3), mean) # rows are topics and columns are ASVs
mean.beta.c = t(mean.beta.c) %>% data.frame
mean.beta.c = mutate(mean.beta.c,
  ASV = taxa_names(ps.c)
)

```


### Posterior distirbution of theta in each control
```{r}
theta.c = samples.c$theta
mean.theta.c = apply(theta.c, c(2,3), mean)
rownames(mean.theta.c) = sample_names(ps.c)
```


## LDA on specimen samples

```{r}
ps = psPlByBlock[[1]] # all controls
x <- t(get_taxa(ps))
K = 10 # number of topics
dimnames(x) <- NULL
stan_data <- list(
  K = K,
  V = ncol(x),
  D = nrow(x),
  n = x,
  alpha = rep(1, K),
  gamma = rep(0.5, ncol(x))
)
```

```{r}
start_fit <- Sys.time()

f <- stan_model(file = "lda.stan")

stan_fit <- vb(
  f,
  data = stan_data,
  output_samples = 1000,
  eta = 1,
  adapt_engaged = FALSE
)

cat(sprintf(
  "Finished in %f minutes\n",
  Sys.time() - start_fit, 4)
)

saveRDS(stan_fit, file = "stan_vb_fit.rds")
```

```{r}
samples <- rstan::extract(stan_fit)
```

### Posterior distirbution of beta in specimen-samples (beta for each topic - not for each topic in each sample)

```{r}
beta = samples$beta

mean.beta = apply(beta, c(2,3), mean) # rows are topics and columns are ASVs
mean.beta = t(mean.beta) %>% data.frame
mean.beta = mutate(mean.beta,
  ASV = taxa_names(ps)
)

```

### Posterior distirbution of theta in each specimen sample
```{r}
theta = samples$theta
mean.theta = apply(theta, c(2,3), mean)
rownames(mean.theta) = sample_names(ps)
```
